src/libstd/sys/unix/process.rs

```diff
@@ -254,6 +259,81 @@ impl Process {
             unsafe { libc::_exit(1) }
         }
 
+        fn walk_open_fd_brute<F>(action_fd: F) where F: Fn(RawFd) {
+            // The `getdtablesize(3)` and `sysconf(_SC_OPEN_MAX)` could miss file descriptors if
+            // the RLIMIT_NOFILE is lowered (cf. #12148) but there is no good way to list them all
+            // except with procfs (cf. `walk_open_fd_proc()`).
+            // However, an unprivileged process (i.e. without CAP_SYS_RESOURCE on Linux) is not
+            // allowed to use a file descriptor upper or equal to it's hard limit nor to change
+            // this limit upwards.
+            let begin = libc::STDERR_FILENO + 1;
+            let end = unsafe { libc::sysconf(libc::consts::os::sysconf::_SC_OPEN_MAX) } as RawFd;
+            for fd in begin..end {
+                action_fd(fd);
+            }
+        }
+
+        #[cfg(any(target_os = "linux", target_os = "android"))]
+        fn walk_open_fd_proc<F>(action_fd: F) where F: Fn(RawFd) {
+            // Take care to look at a real procfs
+            let is_procfs = match ::fs::metadata("/proc") {
+                Ok(fs) => {
+                    // The device ID is 3 for a bare Linux but change for other namespaces (e.g.
+                    // with LXC). With Linux user namespace, the UID and GID may be different than
+                    // 0, so we only check the root inode.
+                    fs.is_dir() && fs.ino() == 1
```

Comment:
I known well sandboxing. To be compatible with absent `/proc`, my patch fall back to the brute force FDs closing.